addons/stock/tests/test_move2.py

```diff
@@ -1810,6 +1810,73 @@ def test_empty_moves_validation_2(self):
         back_order = self.env['stock.picking'].search([('backorder_id', '=', delivery_order.id)])
         self.assertFalse(back_order, 'There should be no back order')
 
+    def test_pick_cancel_with_moves(self):
+
+        move = self.MoveObj.create({
+            'name': self.productA.name,
+            'product_id': self.productA.id,
+            'product_uom_qty': 10,
+            'product_uom': self.productA.uom_id.id,
+            'location_id': self.pack_location,
+            'location_dest_id': self.customer_location,
+        })
+
+        # create picking with stock move.
+        picking = self.env['stock.picking'].create({
+            'location_id': self.pack_location,
+            'location_dest_id': self.customer_location,
+            'partner_id': self.partner_delta_id,
+            'move_ids_without_package':move,
+            'picking_type_id': self.picking_type_out,
+            })
+
+        picking.action_confirm()
+        self.assertEqual(picking.state, 'confirmed', "Picking should be in a confirmed state.")
+
+        # picking in a confirmed state and try to cancel it.
+        picking.action_cancel()
+        self.assertEqual(picking.state, 'cancel', "Picking should be in a cancel state.")
+
+    def test_pick_cancel_without_moves(self):
+
+        # create picking wihtout stock move but with move_line and immediate transfer is 1.
+        picking = self.env['stock.picking'].create({
+            'location_id': self.pack_location,
+            'location_dest_id': self.customer_location,
+            'partner_id': self.partner_delta_id,
+            'picking_type_id': self.picking_type_out,
+            'immediate_transfer': True,
+            })
+
+        move_line = self.StockPackObj.create({
```

Comment:
add a comment before:
"On this immediate transfer, add a move line and a package level"